File: python/game.py
import sys
from grid import Grid
import time

if __name__ == '__main__':
    input = sys.argv[1]
    output = sys.argv[2]
    iterations = int(sys.argv[3])
    interactive = len(sys.argv) > 4

    width = 0
    height = 0

    with open(input) as f:
        for line in f:
            if (width == 0):
                width = len(line) // 2
            if line:
                height += 1

    grid = Grid(width, height)

    with open(input) as f:
        y = 0
        for l in f:
            x = 0
            for i, c in enumerate(l):
                if i % 2 == 0:
                    alive = c == 'X'
                    grid.set_alive(x, y, alive)
                    x += 1
            y += 1

    grid2 = Grid(width, height)
    for _ in range(iterations):
        grid.step_all(grid2)
        grid, grid2 = grid2, grid
        if interactive:
            print(chr(27) + "[2J")
            print(grid)
            time.sleep(0.1)

    with open(output, 'w') as f:
        for y in range(grid.height):
            for x in range(grid.width):
                a = grid.get_alive(x, y)
                f.write('X' if a else '.')
                f.write(' ')
            f.write('\n')

File: java/Game.java
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;

public class Game {
    public static void main(String[] args) throws IOException, InterruptedException {
        String input = args[0];
        String output = args[1];
        long iterations = Long.parseLong(args[2]);
        boolean interactive = args.length > 3;

        Grid grid = readGrid(input);
        iterateGrid(grid, iterations, interactive);
        writeGrid(grid, output);
    }

    public static Grid iterateGrid(Grid grid, long iterations, boolean interactive) {
        Grid grid2 = new Grid(grid.width, grid.height);

        for (long i = 0; i < iterations; i++) {
            grid.stepAll(grid2);

            Grid tmp = grid;
            grid = grid2;
            grid2 = tmp;

            if (interactive) {
                System.out.print((char) 27);
                System.out.print("[2J");
                System.out.println(grid);
                //Thread.sleep(100);
            }
        }

        return grid;
    }

    public static Grid readGrid(String filepath) throws IOException {
        int width = 0;
        int height = 0;

        try (var reader = Files.newBufferedReader(new File(filepath).toPath())) {
            while (reader.ready()) {
                String line = reader.readLine();
                if (width == 0) {
                    width = (int) Math.ceil(line.length() / 2.0);
                }
                if (!line.isEmpty()) {
                    height++;
                }
            }
        }

        Grid grid = new Grid(width, height);

        try (var reader = Files.newBufferedReader(new File(filepath).toPath())) {
            int y = 0;
            while (reader.ready()) {
                String line = reader.readLine();
                for (int x = 0; x < width; x++) {
                    char c = line.charAt(x * 2);
                    boolean alive = c == 'X';
                    grid.setAlive(x, y, alive);
                }
                y++;
            }
        }

        return grid;
    }

    public static void writeGrid(Grid grid, String filepath) throws IOException {
        try (var writer = Files.newBufferedWriter(new File(filepath).toPath())) {
            for (int y = 0; y < grid.height; y++) {
                for (int x = 0; x < grid.width; x++) {
                    boolean a = grid.getAlive(x, y);
                    writer.write(a ? 'X' : '.');
                    writer.write(' ');
                }
                writer.write('\n');
            }
        }
    }
}
